Support for jsonformat in duration deserializer

obarcelonap opened the pull request:
Implementation based on Duration::of(long,TemporalUnit). Closes #184

kupci commented on datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/DurationDeserializer.java:
```diff
@@ -79,20 +88,38 @@ protected DurationDeserializer withLeniency(Boolean leniency) {
+                if (pattern != null) {
+                    return pattern.parse(intValue);
+                }
+                return _fromTimestamp(context, intValue);
```
So the new functionality will override the flag `DeserializationFeature.READ_DATE_TIMESTAMPS_AS_NANOSECONDS` if already set... guessing that is ok, though to be safer you might want to adhere to the flag if both are set, and there have a conflict, i.e they don't specify nanos in the pattern. The documentation should be updated too.

obarcelonap replied:
I think not honouring the flag is ok since the `@JsonFormat` annotation has been used explicitly may have more precedence. Actually that flag is not well used in this serializer because it uses `ofSeconds` while the flag is about nanos.

obarcelonap replied:
BTW, what docs should I update?

kupci replied:
I think a line here, to add Duration to the list of exceptions, and mention that it supports "pattern", would be helpful (in the section after the line below)

> Some exceptions to this standard serialization/deserialization rule:

https://github.com/FasterXML/jackson-modules-java8/tree/master/datetime

obarcelonap replied:
Done

kupci commented on datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/DurationDeserializer.java:
```diff
@@ -121,13 +148,18 @@ protected Duration _fromString(JsonParser parser, DeserializationContext ctxt,
         }
 
+        if (pattern != null) {
+            return pattern.parse(NumberInput.parseLong(value));
```
Exception handling tested?

obarcelonap replied:
I generalized the handling of `DateTimeException` in the base class and tested the code.

kupci commented on datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/DurationDeserializer.java:
```diff
@@ -49,6 +53,11 @@
+    /**
+     * Since 2.12
+     */
+    private DurationPattern pattern;
```
Recommend adding some JavaDoc here to describe the new variable. Also suggest naming it `_durationPattern` for consistency with other private vars and to be a little more specific.

obarcelonap replied:
Done

kupci reviewed (commented):
Took a brief look, looks good, may have some thoughts later. One thing that will be nice about this change is that it looks like .._eventually_ one of the DeserializationFeature flags can be deprecated.

kupci commented on datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/DurationDeserializer.java:
```diff
@@ -79,20 +88,38 @@ protected DurationDeserializer withLeniency(Boolean leniency) {
             }
+            if (format.hasPattern()) {
+                this.pattern = DurationPattern.from(format.getPattern())
+                        .orElse(null);
```
Wondering if this should follow the "pattern" of other configuration settings, and set pattern in the super class, and then call a method like the following:
`deser = deser.withPattern(format.getPattern());`

This would set the groundwork for other types adding support for 'pattern's, which might help with some tricky bugs. Or is this duplicating work in date-time-config? Thoughts @cowtowncoder ?

obarcelonap replied:
That could make sense, but I think is a bigger topic so I Ieft the disussion up to you :)

obarcelonap replied:
But in any case, adding a unit in JsonFormat is not a conflict with the current change and it can be added later on.

obarcelonap commented:
Did the changes in form of `--fixup` I will rebase once is ok from your side

kupci reviewed (commented):
Thanks for the changes, they look good to me. Approved - assuming you'll get the documentation in. If they look good to @cowtowncoder too, you'll need to send him the CLA if you haven't already.

kupci commented on datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/DurationDeserializer.java:
```diff
@@ -49,6 +53,11 @@
     public static final DurationDeserializer INSTANCE = new DurationDeserializer();
 
+    /**
+     * Since 2.12
```
Suggest adding complete JavaDoc, see this for a template you could use: https://github.com/FasterXML/jackson-modules-java8/blob/eeb2b67de6fdc975d9bde88d32b2c1c83a18d710/datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/JSR310DateTimeDeserializerBase.java#L22

obarcelonap commented:
> Thanks for the changes, they look good to me. Approved - assuming you'll get the documentation in. If they look good to @cowtowncoder too, you'll need to send him the CLA if you haven't already.

I haven't sent the CLA, where can I find it?

Just rebased on top of 2.12 branch, added missing documentation and squashed all the commits.

cowtowncoder commented:
CLA can be found from `jackson` repo:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

obarcelonap commented:
Just sent the CLA by email.

kupci commented on datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/DurationDeserializer.java:
```diff
@@ -129,9 +164,13 @@ protected Duration _fromString(JsonParser parser, DeserializationContext ctxt,
-        } catch (DateTimeException e) {
-            return _handleDateTimeException(ctxt, e, value);
+        } catch (NumberFormatException | DateTimeException e) {
+            return _handleWeirdStringValue(ctxt, e, value);
```
@cowtowncoder  Ok to use same exception handler method, for a number? Or should be _handleWeirdNumberValue(...)?

obarcelonap replied:
I think is ok to use same handler because at the end the original value that is on processing is a `String`, so it matches the handler purpose.

obarcelonap replied:
since `String` is not handled anymore I reverted these changes.

kupci commented on datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/DurationDeserializer.java:
```diff
@@ -79,20 +97,36 @@ protected DurationDeserializer withLeniency(Boolean leniency) {
                 }
             }
+            if (format.hasPattern()) {
+                deser = DurationUnitParser.from(format.getPattern())
```
@obarcelonap Are you sure this preserves the `leniency` setting from above? Or should follow same convention as above, i.e.
`deser = deser.withPattern(..) `

obarcelonap replied:
Yes it is, `withPattern` is referenced when there is a value using the `Optional` monad.
As you already pointed, the non functional approach, would be:
```
deser = deser.withPattern(DurationUnitParser.from(format.getPattern()).orElse(null))
```
Let me know if you prefer this second approach.

kupci replied:
> would it not be better to simply throw an exception to let user know that fractions are not allowed for anything but seconds?

Didn't even think of that, was assuming no fractions. So yes, my vote would be "simpler is better".

kupci reviewed (commented):
Added a couple more comments/2nd (or third) thoughts, since figuring easier to fix here than later..

obarcelonap reviewed (commented):
> Added a couple more comments/2nd (or third) thoughts, since figuring easier to fix here than later..

Just replied, I think is mainly code style but I can change the code if it fits better in this repo codebase. Let me know

cowtowncoder commented:
Oh. Hmm. I should have looked at this earlier. I'll have to think how I feel about specifying new patterns -- but it is good that at least it is based on `ChronoUnit`; and the limitation of Java annotations prevents doing much more than matching it explicitly.
I'll go over code and hopefully will be able to merge this soon.

cowtowncoder commented on datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/DurationDeserializer.java:
```diff
@@ -79,20 +97,36 @@ protected DurationDeserializer withLeniency(Boolean leniency) {
-                return DecimalUtils.extractSecondsAndNanos(value, Duration::ofSeconds);
+                BigDecimal decValue = parser.getDecimalValue();
+                if (_durationUnitParser != null) {
+                    return _durationUnitParser.parse(decValue.longValue());
```
This would be lossy... I wonder if the right handling would be throw an exception; or perhaps require a coercion check? (2.12 has `CoercionConfig` that could be checked to see if "floating-point -> integer" coercion is legal; by default it is, but can prevent.
This assuming that it is not possible to use fractions in meaningul way... is there?

kupci replied:
Limit the DurationParser to only `ID_NUMBER_INT` logic, to match the longs that are available in the `toXXX` methods of `Duration`, perhaps? (Now that I'm thinking about how this would be serialized.)
https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html#toDays--

cowtowncoder replied:
Yeah; it is easy enough to add support for `String` coercion if need be (if requested).
So let's go with that.

I might still go ahead and check for `CoercionConfig` for existing coercions -- I know "lenient" is used in some places, but I think my thinking has evolved so that there is difference between:

1. Coercion from different input value shape (like JSON String) into something that would usually be required (like integral number). For this we have CoercionConfig and a few legacy features. Leniency does not come to play as it is too coarse setting (on/off)
2. Leniency would affect handling within expected value domain -- so, for String representation of Date, whether we allow "2021-02-30", or similar (case-sensitivity option exists, too, which is yet separate)

kupci replied:
So leave this case block ( case JsonTokenId.ID_NUMBER_FLOAT:) as-is, with no changes, and the only case block that should be changed with this PR is the one for `JsonTokenId.ID_NUMBER_INT'? Keep things simpler for now?

cowtowncoder replied:
Well, kind of; change should be isolated in `_fromTimestamp()`, which can also be called from `_fromString()` -- for special case of XML, CSV etc (formats for which `ctxt.isEnabled(StreamReadCapability.UNTYPED_SCALARS)` returns true).
This because those formats can not reliably indicate number type (XML not without Schema, CSV not without something external), so auto-detection of integral numbers makes sense without coercion.

obarcelonap replied:
Only integers are handled now

cowtowncoder commented on datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/JSR310DeserializerBase.java:
```diff
@@ -141,6 +141,13 @@ protected boolean _isValidTimestampString(String str) {
+    }
+
+    @SuppressWarnings("unchecked")
+    protected <R, E extends RuntimeException> R _handleWeirdStringValue(DeserializationContext context,
```
No need to add type parameter `E`, I think, ok to just pass `Exception`, I think.

cowtowncoder commented:
Ok I think code looks pretty good on its own; I can make small changes myself if need be.

So the first question I have is whether decimal/floating-point numbers (vs integers) should be handled -- and it looks like only "ofSeconds()" handles two parts, although in theory we could calculate "1.5 hours" into 90 minutes just fine... but it gets complicated and I wonder would it not be better to simply throw an exception to let user know that fractions are not allowed for anything but seconds? 

Second part is wrt serialization: should format not also be considered when serializing?

cowtowncoder commented on datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/DurationDeserializer.java:
```diff
@@ -141,4 +180,23 @@ protected Duration _fromTimestamp(DeserializationContext ctxt, long ts) {
+
+        static Optional<DurationUnitParser> from(String unit) {
+            return Stream.of(ChronoUnit.values())
+                    .filter(u -> u.toString().equalsIgnoreCase(unit))
```
I think we should not actually allow case-insensitive match, but require upper-case, to exactly match `ChronoUnit`.

obarcelonap replied:
Now matching is case sensitive.

cowtowncoder commented on datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/DurationDeserializer.java:
```diff
@@ -129,9 +164,13 @@ protected Duration _fromString(JsonParser parser, DeserializationContext ctxt,
 
         try {
+            if (_durationUnitParser != null) {
+                return _durationUnitParser.parse(NumberInput.parseLong(value));
```
Ok, here's another potential question: should JSON Strings be allowed with pattern?
In that case we'd probably want to check `CoercionConfig` as input absolutely must be (integer) number, so finding JSON String requires coercion. Same is not true for default `Duration` which conversely must get String.

I can probably play with this a bit so above is just my thinking out aloud.

obarcelonap replied:
Only integers respects pattern for the moment.

kupci commented on datetime/src/test/java/com/fasterxml/jackson/datatype/jsr310/deser/DurationUnitParserTest.java:
```diff
@@ -0,0 +1,77 @@
+                asArray("Centuries", ChronoUnit.CENTURIES),
+                asArray("Millennia", ChronoUnit.MILLENNIA),
+                asArray("Eras", ChronoUnit.ERAS),
+                asArray("Forever", ChronoUnit.FOREVER)
```
TIL: There is a `ChronoUnit` for "forever". Wondering if we can trim down the actual realistically supported `ChronoUnit`s?

obarcelonap replied:
Not supported, fixed. Actually everthing greater than `DAYS` is not supported because is considered not precise. I fixed the code and added some tests.

kupci commented:
> Second part is wrt serialization: should format not also be considered when serializing?

Hmm, glossed over this, but now I think so. The `DurationSerializer` supports milliseconds, seconds/nanoseconds, or ["string representation"](https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html#toString--):

> A string representation of this duration using ISO-8601 seconds based representation, such as PT8H6M12.345S. 

Now in reviewing, it looks like we can trim supported units down the actual permitted `ChronoUnit`s supported by `Duration` methods, i.e. `toDays()`, `toHours()`, `toMillis()`, `toMinutes()`, `toNanos()`? 

Probably also need to test (with the serializer) edge cases like `ChronUnit.FOREVER` ;)

cowtowncoder commented:
I think I have a few different things/ideas to discuss: will add to issue itself. I hope we can get this merged soon, but I really want to Do It Right, including backwards compatibility.

obarcelonap commented:
Solved your comments guys, thanks for your feedback!
